File: src/abc032/c.py

# テストケース
n = 8
k = 78
s = [9,8,5,6,2,9,1,2]

# 愚直な解
for l in range(n):
    temp = 1
    for r in range(l, n):
        temp *= s[r]
        if temp > k:
            print("l = {}, ans = {}, tmep = {}".format(l,  r-l, temp / s[r]))
            break
        if r == n - 1:
            print("l = {}, ans = {}, tmep = {}".format(l,  1+r-l, temp))


# テストケース
# 8 78
# 9 <- 0
# 8 <- 1
# 5 <- 2
# 6 <- 3
# 2 <- 4 *max len = 4*
# 9 <- 5
# 1 <- 6
# 2 <- 7
# 答えは4

# l = 0, ans = 2, tmep = 72.0
# l = 1, ans = 2, tmep = 40.0
# l = 2, ans = 3, tmep = 60.0
# l = 3, ans = 2, tmep = 12.0
# l = 4, ans = 4, tmep = 36
# l = 5, ans = 3, tmep = 18
# l = 6, ans = 2, tmep = 2
# l = 7, ans = 1, tmep = 2

File: src/abc032/c.rs
#![allow(dead_code, unused_imports)]
use proconio::{input, marker::{Usize1, Isize1}};
use itertools::Itertools;
use std::cmp::{max, min};
use std::cmp::Ordering;
use std::collections::VecDeque;
use std::collections::BinaryHeap;
use std::collections::HashMap;
use std::collections::BTreeMap;
use std::ops::Bound::{Excluded, Included, Unbounded};
use std::collections::{HashSet, BTreeSet};
use proconio::marker::Chars;
use std::f64::consts::PI;
use std::mem::swap;
use superslice::*;
use rand::Rng;
fn main() {
    // 2024-12-13 19:50-20:23 (33min)
    input! {
        n: usize,
        k: usize,
        s: [usize; n],
    }
    // 要素の積がk以下

    let mut is_all_over_k = true;
    for i in 0..n {
        if s[i] <= k {
            is_all_over_k = false;
        }
        if s[i] == 0 {
            // 0 が一個でもあれば、Nにできる。
            println!("{}", n);
            return;
        }
    }
    if is_all_over_k {
        println!("0");
        return;
    }

    // 累積積は使っちゃいけなさそう。
    // 尺取かな

    let mut ans = 1;
    let mut deque = VecDeque::new();
    let mut p = 1;
    let mut min_r = 0;
    for l in 0..n {
        // println!("l = {:?} -------", l);
        loop {
            if let Some(&v) = deque.front() {
                if v < l {
                    let v = deque.pop_front().unwrap();
                    p /= s[v];
                }
                else if v == l {break}
            }
            else {
                // 無ければ追加
                deque.push_back(l);
                p *= s[l];
                break
            }
        }
        if p > k {continue}
        min_r = max(l, min_r);

        // println!("deque = {:?}", deque);
        // println!("min_r = {:?}", min_r);
        // println!("p = {:?}", p);

        // 先頭が l な状況は作った。
        for r in min_r+1..n {
            min_r = r;
            p *= s[r];
            deque.push_back(r);
            if p > k {
                p /= s[r];
                // println!("p = {:?}", p);
                min_r = r-1;
                deque.pop_back();
                break
            }
        }
        ans = max(ans, deque.len());
    }
    println!("{}", ans);
}